packages/smithy-core/src/smithy_core/documents.py

```diff
@@ -640,22 +640,53 @@ def read_timestamp(self, schema: "Schema") -> datetime.datetime:
         return self._value.as_timestamp()
 
 
-# A registry for on-demand deserialization of types by using a mapping of shape IDs to their deserializers.
 class TypeRegistry:
+    """A registry for on-demand deserialization of types by using a mapping of shape IDs
+    to their deserializers."""
+
     def __init__(
         self,
         types: dict[ShapeID, type[DeserializeableShape]],
         sub_registry: "TypeRegistry | None" = None,
     ):
+        """Initialize a TypeRegistry.
+
+        :param types: A mapping of ShapeID to the shapes they deserialize to.
+        :param sub_registry: A registry to delegate to if an ID is not found in types.
+        """
         self._types = types
         self._sub_registry = sub_registry
 
     def get(self, shape: ShapeID) -> type[DeserializeableShape]:
+        """Get the deserializable shape for the given shape ID.
+
+        :param shape: The shape ID to get from the registry.
+        :returns: The corresponding deserializable shape.
+        :raises KeyError: If the shape ID is not found in the registry.
+        """
         if shape in self._types:
             return self._types[shape]
         if self._sub_registry is not None:
             return self._sub_registry.get(shape)
         raise KeyError(f"Unknown shape: {shape}")
 
+    def __getitem__(self, shape: ShapeID):
+        """Get the deserializable shape for the given shape ID.
+
+        :param shape: The shape ID to get from the registry.
+        :returns: The corresponding deserializable shape.
+        :raises KeyError: If the shape ID is not found in the registry.
+        """
+        return self.get(shape)
+
+    def __contains__(self, shape: ShapeID):
```

Comment:
```suggestion
    def __contains__(self, item: object, /):
```

I really wish type checkers would check this. Dunder methods in general aren't well supported and I don't know why. It's very easy to get this wrong. In fact we only do it right in 2 of 6 usages. We also have one `Any` which is actually more wrong. And I'm almost certainly guilty of all or nearly all of those error cases.

In this particular case the implementation doesn't rely on the given item being a `ShapeID` so there's no actual runtime issue.

Reply:
Oh interesting, thanks for the explanation too.